src/libfsm/gen.c

```diff
@@ -145,6 +145,23 @@ fsm_generate_matches(struct fsm *fsm, size_t max_length,
 	return res;
 }
 
+enum fsm_generate_matches_cb_res
+fsm_generate_cb_printf(const struct fsm *fsm,
+	size_t depth, size_t match_count, size_t steps,
+	const char *input, size_t input_length,
+	fsm_state_t end_state, void *opaque)
+{
+	(void)fsm;
+	(void)input_length;
+	(void)end_state;
+	(void)opaque;
+	(void)depth;
+	(void)match_count;
+	(void)steps;
+	printf("%s\n", input);
```

Comment:
We're defining a line-delimited protocol with the `\n` here, so the output isn't binary. Let's escape non-printable characters (use `c_escputc_str()` for that). And the CLI tools need to heed `-X` for "always_hex" output, so I suppose that has to be an additional callback.

Reply:
Changed in 8219d433.

Shoulds there a way to fully _disable_ escaping in the CLI?

Reply:
No - I don't see how we ever could disable escaping, because we're printing meaningful newlines and such, so this is always a kind of protocol

Reply:
That is, i don't think an unescaped `fsm_generate_cb_printf()` ever makes sense

Reply:
Okay. I didn't want escaping when comparing results with PCRE, but that wasn't using the CLI interface at all, and as you said that's inherently a line-oriented protocol.

Reply:
Oh I see, we don't need an additional callback because `c_escputc_str()` takes an `opt`. That's fab, thank you!